Support SET SCHEMA and SET SEARCH_PATH

omo opened the pull request:
Please take a look when you have time?
This series of changes fix #155.

I found some interesting things along the way. These can be addressed in separate PR:

 * The "temp" schema is a bit limiting. We might want to support TEMP tables without using this?
 * Using "main" schema for built-ins can be problematic because since these built-ins become invisible after
   setting the search_path. We might give them their own, like "builtin" or something.
 * The queries in CREATE VIEW can break after the search_path change. We might want to either
   pre-compile these queries somehow or retain the search path for the later resolution.

Anyway, these aren't the main points of this PR. 

Also, I credit #1451 where I borrowed the test. Thanks @rohankumar42!

omo committed:
Inroduce CatalogSearchPath.

This change introduces CatalogSearchPath class.
The class lazily populates the search path list from
the 'search_path' setting value and replaces the string list for
ClientContext::catalog_search_path.

The SET and CURRENT_SET is also updated for handling validation
and aliasing of SEARCH_PATH and SCHEMA set variables.

The search order is now:
  "temp" -> given search paths -> "main" -> "pg_catalog".

The side effect is yet to be thoroughly tested and probably broken.
Following changes will take care of it.

omo committed:
SET and CURRENT_SETTING: Normalize the keys to lowercase.

This is necessary as some setting values are going to have
specific semantics.

Preparing for #155.

omo committed:
Make basic lookup work.

This change actually starts using CatalogSearchPath for the lookup.
Some things are still broken and need mor work, as the test TODO suggests.
A notable imcommpatibility is built-in function name lookup.
We put them into "main", but it needs its own to be always-visible.

Tests are borrowed from #1451. Kudos @rohankumar42!

omo committed:
Make CREATE VIEW mostly work.

Note that this change reveals a slight incompatibility
between Duck and PG, where PG binds the table on CREATE VIEW,
while Duck binds it on SELECT. Although the PG behavior seems
more legit, this change leaves it as is for now.

omo committed:
Make CREATE MACRO and CREATE SEQUENCE work.

Actually these have worked even without this change
since the "schema" member varaible is replaced in all the places.
Fixing CreateMacroInfo and CreateSequenceInfo just to make them future-change proof.

omo committed:
Make ALTER and DROP work.

This change involves some refactoring around catalog entry lookup:
It adds Catalog::LookupEntry() to return both the entry and its
associated scheme. This simplifies some codepaths.

As the side effect, it now also handles DROP TABLE IF EXISTS
against a table on a missing schema, which should be allowed but wasn't.

omo commented:
Tried to run SQLancer locally but no luck :-(

Mytherin commented on src/execution/operator/helper/physical_set.cpp:
```diff
@@ -7,13 +9,39 @@ namespace duckdb {
+
+		try {
+			for (const auto &path : paths) {
+				context.client.db->GetCatalog().GetSchema(context.client, StringUtil::Lower(path));
```
Can we do this without the try/catch? We try to avoid try/catch as much as possible outside of central locations (e.g. the client context, etc).

omo replied:
Done.

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some comments below.

> The "temp" schema is a bit limiting. We might want to support TEMP tables without using this?

I'm not opposed to letting temporary tables live in any schema. The main point of the temporary schema is that this is connection-local, so temporary objects from one connection do not interfere with temporary objects from another connection. We could let this be the default for temporary objects, and if the schema is explicitly specified elsewhere store them in that schema. 

> Using "main" schema for built-ins can be problematic because since these built-ins become invisible after setting the search_path. We might give them their own, like "builtin" or something.

The easiest solution would be to always include the "main" schema on the search path (similar to how postgres always includes pg_catalog on the search path). We could consider adding a separate SYSTEM_SCHEMA (e.g. `duckdb`) that holds the built-in functions like pg_catalog. 

> The queries in CREATE VIEW can break after the search_path change. We might want to either pre-compile these queries somehow or retain the search path for the later resolution.

Yes, that is a good point. One possible solution here would be to store the search path with the view, and then on the binder level switch the active search path around. Not super crucial for an initial pass.

omo commented on src/main/client_context.cpp:
```diff
@@ -936,4 +939,9 @@ bool ClientContext::TryGetCurrentSetting(const std::string &key, Value &result)
 }
 
+const std::exception &ClientContext::SetLastError(const std::exception &error) {
+	set_variables["__last_error"] = Value(error.what());
```
Ah makes sense. Thanks for the catch!
I dropped this commit from the PR for now.

omo commented on src/catalog/catalog_search_path.cpp:
```diff
@@ -0,0 +1,48 @@
+			paths.emplace_back(StringUtil::Lower(p));
+		}
+	} else {
+		paths.emplace_back(DEFAULT_SCHEMA);
```
Done.

omo commented on src/include/duckdb/catalog/catalog_search_path.hpp:
```diff
@@ -0,0 +1,37 @@
+//! The schema search path, in order by which entries are searched if no schema entry is provided
+class CatalogSearchPath {
+public:
+	typedef std::function<bool(Value &)> ValueGetter;
```
Okay, I was hesitated to spread ClientContext dependency but this was probably overkill.

omo commented:
Thanks for the review Mark! PTAL?
The update
 * dropped the error message retrieval and "did you mean" handling. I'll send this part as separate PR once this is in.
 * added DEFAULT_SCHEMA back to the search path regardless of the user-set schema.

----
> I'm not opposed to letting temporary tables live in any schema. The main point of the temporary schema is that this is connection-local, so temporary objects from one connection do not interfere with temporary objects from another connection. We could let this be the default for temporary objects, and if the schema is explicitly specified elsewhere store them in that schema.
> 
Oh I didn't think about this. Thanks for the catch!
Although it would be possible to extend schema / catalog entries to be connection-aware, it might be a bit too complicated for its value.

> > Using "main" schema for built-ins can be problematic because since these built-ins become invisible after setting the search_path. We might give them their own, like "builtin" or something.
> 
> The easiest solution would be to always include the "main" schema on the search path (similar to how postgres always includes pg_catalog on the search path). We could consider adding a separate SYSTEM_SCHEMA (e.g. `duckdb`) that holds the built-in functions like pg_catalog.
> 
This would be nice. Then we can drop DEFAULT_SCHEMA from the search path, which I guess what PG does.
But let me leave DEFAULT_SCHEMA in the search path for now. Adding SYSTEM_SCHEMA will be a big change.

> > The queries in CREATE VIEW can break after the search_path change. We might want to either pre-compile these queries somehow or retain the search path for the later resolution.
> 
> Yes, that is a good point. One possible solution here would be to store the search path with the view, and then on the binder level switch the active search path around. Not super crucial for an initial pass.

Agreed. Leaving it as is.

omo commented on src/include/duckdb/main/client_context.hpp:
```diff
@@ -213,6 +213,9 @@ class ClientContext : public std::enable_shared_from_this<ClientContext> {
 	bool UpdateFunctionInfoFromEntry(ScalarFunctionCatalogEntry *existing_function, CreateScalarFunctionInfo *new_info);
 
+	//! Remember the last exception.
+	const std::exception &SetLastError(const std::exception &error);
```
Will do.

Mytherin commented on src/parser/transform/statement/transform_set.cpp:
```diff
@@ -36,7 +38,7 @@ unique_ptr<SetStatement> Transformer::TransformSet(duckdb_libpgquery::PGNode *no
 	}
 
-	auto name = std::string(stmt->name);
+	auto name = StringUtil::Lower(std::string(stmt->name));
```
Rather than remembering to use `StringUtil::Lower` in all locations where SET variables are used, could we turn the `set_variables` into a `case_insensitive_map`?

omo replied:
Oh wasn't aware of this. Thanks for the tip!

Mytherin commented on src/function/scalar/generic/current_setting.cpp:
```diff
@@ -24,6 +26,17 @@ static void CurrentSettingFunction(DataChunk &args, ExpressionState &state, Vect
 }
 
+static string NormalizeSettingName(const string &name) {
+	auto lower = StringUtil::Lower(name);
```
Same here, explicit `StringUtil::Lower` should not be required if we are using the `case_insensitive_map`.

Mytherin commented on src/execution/operator/helper/physical_set.cpp:
```diff
@@ -7,13 +9,37 @@ namespace duckdb {
+		}
+
+		for (const auto &path : paths) {
+			if (!context.client.db->GetCatalog().GetSchema(context.client, StringUtil::Lower(path), true)) {
```
Explicit `StringUtil::Lower` should not be required here as the catalog uses case insensitive lookups.

Mytherin commented on test/sql/catalog/test_set_search_path.test:
```diff
@@ -0,0 +1,110 @@
+SET SCHEMA = 'test';
+
+statement error
+SET SCHEMA = 'test,test2';
```
Should this give an error? Remember that schemas can contain commas (although this might be ill-advised!).

omo replied:
Oh good point! Here is what I found:

 * PG supports quoting like`SET SEARCH_PATH '"foo","bar,baz."'` although it doesn't support escape. I added this too.
 * SET SCHEMA is deprecated and now gone in the latest PG. I'd leave the support in duckdb though as people seem like it.
 * Apparently our CREATE SCHEMA doesn't support comma-included names so I couldn't write  SQL test. The logic is covered by the unit tests so I hope it is sufficient.

Mytherin commented on test/sql/catalog/test_set_search_path.test:
```diff
@@ -0,0 +1,110 @@
+SET SCHEMA = 'test,test2';
+
+statement error
+SET SCHEMA = 'does_not_exist';
```
Could you add a test for `SET SCHEMA` and `SET SEARCH_PATH` case insensitivity?

omo replied:
Done.

Mytherin reviewed (commented):
Thanks for the fixes! This looks a lot better now. Some more comments below.

The node.js failure seems to be related to the error message changing when a table lookup fails:

```sql
-- current
Error: Catalog Error: Table with name foo does not exist!
Did you mean "fow"?
LINE 1: select * from foo;

-- this PR
D select * from foo;
Error: Catalog Error: foo is not found in the search path temp,main,pg_catalog
LINE 1: select * from foo;
```

Could we revert this back to the old error message? The search path is generally not relevant to users and might end up confusing beginners, and giving a suggestion in case of a typo is in my experience very helpful.

A more helpful error relating to the search path might be to search through all schemas for the table, and if a table was found in another schema, we could say something like:

```sql
Catalog Error: Table with name foo does not exist!
Table with name foo was found in schema "myschema", but this schema is not part of the search path "temp,main,pg_catalog".
Did you mean "myschema.foo"?
```

omo committed:
Improve "did you mean" suggestion.

This CL imporves the "did you mean" suggestion such that:

 * It suggests candidates for non-qualified name. In that case
   the candidates are searched from search_path.
 * It also look up the candidate from non search_path schemas
   so that it can find better one.

The test for the suggested error messages will be done in a
separate PR. See did_you_mean.test.

omo commented:
Okay, the "did you mean" suggestion is back with the suggested improvement.
PTAL?

Mytherin commented on test/sql/catalog/test_set_search_path.test:
```diff
@@ -0,0 +1,141 @@
+
+# Setting the default value.
+statement ok
+SET SEARCH_PATH = 'test';
```
Could we add tests for:
* A single quoted value (`"test"`)
* A single entry with a quote in the middle of it (`hello"world`)
* Many entries (e.g. 5 entries)
* Various empty entries (e.g. the following strings: `''`  `',,'` `'""'` `'"",""'` `'hello,""'`)
* Duplicate entries in the search path (`main,main`) (should this be allowed?)
* A test with multiple connections that verifies setting the SEARCH_PATH is local to each connection (see e.g. `test/sql/transactions/test_index_pending_update.test` for a test that uses multiple connections)

omo replied:
> Could we add tests for:
> 
> * A single quoted value (`"test"`)
> * A single entry with a quote in the middle of it (`hello"world`)
> * Various empty entries (e.g. the following strings: `''`  `',,'` `'""'` `'"",""'` `'hello,""'`)
These are (already/newly) added to the C++ unit test. It's hard to test these from SQL since it's not clear if the code path is correctly exercised or if it is rejected due to the missing schema name.

> * Many entries (e.g. 5 entries)
> * Duplicate entries in the search path (`main,main`) (should this be allowed?)
> * A test with multiple connections that verifies setting the SEARCH_PATH is local to each connection (see e.g. 
Done. I thinks having duplicate is fine. The semantics is clear enough.

Mytherin reviewed (commented):
Thanks for the changes! This is looking great. Big fan of the new error messages!

Some more minor comments, and then I think it's ready to merge:

omo commented on src/catalog/catalog.cpp:
```diff
@@ -174,29 +177,100 @@ void Catalog::ScanSchemas(ClientContext &context, std::function<void(CatalogEntr
-		// no schema provided: check the catalog search path in order
-		if (context.catalog_search_path.empty()) {
-			throw InternalException("Empty catalog search path");
+std::tuple<string, idx_t, SchemaCatalogEntry *>
```
Yep. The code was a bit too sloppy :-(

Mytherin commented:
Thanks for the changes! Looks good now. Will merge after the tests pass.
